File: blacksmith/Forge.cs
using System;
using System.Collections.Generic;

class Forge {
    public string name;
    public List<Material> requiredMaterials { set; get; }
    public List<int> itemRatio { set; get; }

    public Forge(string RecipeName){
       this.name = RecipeName;
       this.requiredMaterials = new List<Material>();
       this.itemRatio = new List<int>();
    }

    public void AddMaterials(Material material){
        this.requiredMaterials.Add(material);
        this.requiredMaterials.Sort((x, y) => x.GetName().CompareTo(y.GetName()));
        this.itemRatio.Add(material.GetAmount());
    }

    public bool CheckItem(Material item){
        foreach (Material itemMaterial in this.requiredMaterials){
            if(item.GetName() == itemMaterial.GetName()){
                return true;
            }
        }

        return false;
    }

    public string ComputeQuality(List<Material> materials) {
        materials.Sort((x, y) => x.GetName().CompareTo(y.GetName()));
        int qualityPoints = 0;

        foreach(Material item in this.requiredMaterials){
            qualityPoints = qualityPoints + item.GetQuality();
        }



        if(qualityPoints >= 5000 ){
            return "Legendary";
        }
        else if (qualityPoints >= 3000){
            return "Rare";
        }
        else {
            return "Basic";
        }

    }

    public bool CheckQuantity(List<int> array){
        foreach(int x in array){
            if(x < 1){
                return false;
            }
        }

        return true;
    }

    public int ComputeQuantity(List<Material> materials){
        materials.Sort((x, y) => x.GetName().CompareTo(y.GetName()));
        List<int> itemStack = new List<int>();

        foreach(Material availItem in materials){
            foreach(Material reqItem in this.requiredMaterials){
                if(availItem.GetName() == reqItem.GetName()){
                    itemStack.Add(availItem.GetAmount() + 1);
                }
            }
        }

        int totalItems = 0;

        while(this.CheckQuantity(itemStack)){
            for (int i = 0; i < itemStack.Count; i++) {
                if(!this.CheckQuantity(itemStack)){
                    return totalItems;
                }

                itemStack[i] = (itemStack[i] - this.itemRatio[i]);
           
            };

            if(this.CheckQuantity(itemStack)){
                totalItems+=1;
            }
        }

        return totalItems;
    }

    public String craft(List<Material> materials){
        materials.Sort((x, y) => x.GetName().CompareTo(y.GetName()));

        if(materials.Count < this.requiredMaterials.Count){
            throw new ArgumentException("You lack the amount of materials needed to craft this item");

        }

        foreach(Material item in materials){
            if(!this.CheckItem(item)){
                string itemName = item.GetName();
                throw new ArgumentException($"{itemName} is a wrong material for this recipe");
            }
        }

        int quantity = this.ComputeQuantity(materials);

        if(quantity <= 0){
            throw new ArgumentException("You lack the amount of materials needed to craft this item");

        }

        string qualityText = this.ComputeQuality(materials);

        return $"Successfully crafted a {qualityText} {this.name}";
    }


    

    static void Main(string[] args){
        Forge sword = new Forge("Bread Sword");
        Material bread = new Material("Bread", 5, 10000);
        Material breadMat = new Material("Bread", 5, 10000);
        List<Material> materials = new List<Material>();

        materials.Add(breadMat);
        sword.AddMaterials(bread);
        Console.WriteLine(sword.craft(materials));
  
    }


}

class Material {
    public string name;
    public int amount;
    public int quality;

    public Material(String name, int amount, int quality){
        this.name = name;
        this.amount = amount;
        this.quality = quality;
    }

    public String GetName(){
        return this.name;
    }

    public int GetAmount(){
        return this.amount;
    }

    public int GetQuality(){
        return this.quality;
    }
}

File: blacksmith/Forge.java
import java.util.ArrayList;
import java.util.Comparator;



import java.util.*;

class Forge {
    public String name;
    List<Integer> itemRatio = new ArrayList<Integer>();
    public List<Material> requiredMaterials = new ArrayList<Material>();

    public Forge(String name) {
        this.name = name;

    }

    public void addMaterials(Material material) {
        this.requiredMaterials.add(material);
        Collections.sort(this.requiredMaterials, Material.MatName);

        this.itemRatio.add(material.getAmount());


    }

    public boolean checkItem(Material item) {

        for (Material itemMaterial : this.requiredMaterials) {
            if (item.getName() == itemMaterial.getName()) {
                return true;
            }
        }

        return false;
    }

    public String computeQuality(List<Material> materials) {

        Collections.sort(materials, Material.MatName);
        int qualityPoints = 0;

        for (Material item : materials) {
            qualityPoints = qualityPoints + item.getQuality();
        }

        if (qualityPoints >= 5000) {
            return "Legendary";
        } else if (qualityPoints >= 3000) {
            return "Rare";
        }

        else {
            return "Basic";
        }

    }

    public boolean checkQuantity(List<Integer> array) {

        for (Integer x : array) {
            if (x < 1) {
                return false;
            }
        }
        return true;

    }

    public Integer computeQuantity(List<Material> materials) {
        Collections.sort(materials, Material.MatName);
        List<Integer> itemStack = new ArrayList<Integer>();

        for (Material availItem : materials) {
            for (Material reqItem : this.requiredMaterials) {
                if (availItem.getName() == reqItem.getName()) {
                    itemStack.add(availItem.getAmount() + 1);
                }
            }
        }

   

        Integer totalItems = 0;

        while(this.checkQuantity(itemStack)) {

            // System.out.println(itemStack.size());
            for (int i = 0; i < itemStack.size(); i++) {
              
                if (!this.checkQuantity(itemStack)) {
                    return totalItems;
                }
                
                itemStack.set(i, itemStack.get(i) - this.itemRatio.get(i));
            };

            if (this.checkQuantity(itemStack)) {
                totalItems++;
            }

        }       

        return totalItems;
    }

    public String craft(List<Material> materials){
        Collections.sort(materials, Material.MatName);

        if(materials.size() < this.requiredMaterials.size()){
            throw new IllegalArgumentException("You lack the amount of materials needed to craft this item");
        }

        for(Material item: materials){
            if(!this.checkItem(item)){
                throw new IllegalArgumentException(String.format("%s is a wrong material for this recipe", item.getName()));
            }
        }

        Integer quantity = this.computeQuantity(materials);

        if(quantity <= 0){
            throw new IllegalArgumentException("You lack the amount of materials needed to craft this item");

        }

        String qualityText = this.computeQuality(materials);

        return String.format("Successfully crafted a %s %s", qualityText, this.name);
    }

    // public static void main(String[] args) {
        // List<Material> availMaterial = new ArrayList<Material>();

        // Material stone = new Material("Stone", 3, 0);
        // Material wood = new Material("Wood", 1, 0);
       

        // Material stoneAvail = new Material("Stone", 12, 0);
        // Material woodAvail = new Material("Wood", 50,0);

        // availMaterial.add(stoneAvail);
        // availMaterial.add(woodAvail);


        // Forge sword = new Forge("Sword Cake");

        // sword.addMaterials(stone);
        // sword.addMaterials(wood);


        // System.out.println(sword.computeQuantity(availMaterial));
    // }
}

class Material {
    public String name;
    public Integer amount;
    public Integer quality;

    public Material(String name, Integer amount, Integer quality) {
        this.name = name;
        this.amount = amount;
        this.quality = quality;
    }

    public static Comparator<Material> MatName = new Comparator<Material>() {

        public int compare(Material i1, Material i2) {

            String nameNo1 = i1.getName();
            String nameNo2 = i2.getName();

            /* For ascending order */
            return nameNo1.compareTo(nameNo2);

            /* For descending order */
            // rollno2-rollno1;
        }
    };

    public String getName() {
        return name;
    }

    public Integer getAmount() {
        return amount;
    }

    public Integer getQuality() {
        return quality;
    }

}